core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java

```diff
@@ -117,22 +117,25 @@ private RequestLogBuilder logBuilder() {
     }
 
     @Override
-    public void onRequestTimeoutChange(long newRequestTimeoutMillis) {
-        // Cancel the previously scheduled timeout, if exists.
-        cancelTimeout();
-
-        if (newRequestTimeoutMillis > 0 && state != State.DONE) {
-            // Calculate the amount of time passed since the creation of this subscriber.
-            final long passedTimeMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTimeNanos);
+    public void resetTimeout(long newRequestTimeoutMillis) {
+        if (newRequestTimeoutMillis > 0) {
+            final long passedTimeNanos = System.nanoTime() - startTimeNanos;
+            setDeadline(TimeUnit.MILLISECONDS.toNanos(newRequestTimeoutMillis) - passedTimeNanos);
+        }
+    }
 
-            if (passedTimeMillis < newRequestTimeoutMillis) {
-                timeoutFuture = ctx.channel().eventLoop().schedule(
-                        this::onTimeout,
-                        newRequestTimeoutMillis - passedTimeMillis, TimeUnit.MILLISECONDS);
-            } else {
-                // We went past the dead line set by the new timeout already.
-                onTimeout();
-            }
+    @Override
+    public void setDeadline(long deadlineNanos) {
+        // Cancel the previously scheduled timeout, if exists.
+        if (!cancelTimeout()) {
```

Comment:
You need to make sure a timeout task is not rescheduled when the deadline did not change for efficiency.